File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # in order traverse , time O(n), space O(H) = O(logn)
        prev = -float('inf')

        def in_order(node: TreeNode | None) -> bool:
            nonlocal prev

            if not node:
                return True
            
            if not in_order(node.left):
                return False
            
            if node.val <= prev:
                return False
            
            prev = node.val

            return in_order(node.right)
        
        return in_order(root)

File: Data Structures & Algorithms/is-palindrome/submission-4.go
func isPalindrome(s string) bool {
	left := 0
	right := len(s) - 1

	for left < right {
		if !isAlphanumeric(s[left]) {
			left++
			continue
		}
		if !isAlphanumeric(s[right]) {
			right--
			continue
		}

		if toLower(s[left]) != toLower(s[right]) {
			return false
		}

		left++
		right--
	}

	return true
}

func isAlphanumeric(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || 
	       (ch >= 'A' && ch <= 'Z') || 
	       (ch >= '0' && ch <= '9')
}

func toLower(ch byte) byte {
	if ch >= 'A' && ch <= 'Z' {
		return ch + 32
	}
	return ch
}
